packages/core/src/components/forms/numericStepper.tsx

```diff
@@ -0,0 +1,255 @@
+    }
+
+    private handleIncrementButtonClick = (e: React.MouseEvent<HTMLInputElement>) => {
+        const direction = 1;
+        this.updateValue(direction, e.altKey, e.shiftKey);
+    }
+
+    private handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+        if (this.props.disabled || this.props.readOnly) {
+            return;
+        }
+
+        const { keyCode } = e;
+
+        let direction: number;
+
+        if (keyCode === Keys.ARROW_UP) {
+            direction = 1;
+        } else if (keyCode === Keys.ARROW_DOWN) {
+            direction = -1;
+        } else {
+            return;
+        }
+
+        // we'd like to select the field contents after running the code in this
+        // onKeyDown handler, as a UX nicety. without e.preventDefault, some
+        // hotkeys (e.g. shift + up/down, alt + up/down) will clear the selection,
+        // resulting in an inconsistent or unintuitive experience.
+        e.preventDefault();
+
+        this.updateValue(direction, e.altKey, e.shiftKey);
+    }
+
+    private handleInputChange = (e: React.KeyboardEvent<HTMLInputElement>) => {
+        // the type definition for EventTarget (e.target) doesn't include
+        // the 'value' property out of the box, hence the janky casting
+        const nextValue = (e.target as any).value;
+        this.setState({ shouldSelectAfterUpdate : false, value: nextValue });
+
+        if (this.props.onChange) {
+            this.props.onChange(e);
+        }
+    }
+
+    private updateValue(direction: number, isAltPressed: boolean, isShiftPressed: boolean) {
+        let delta = this.props.stepSize * direction;
+
+        if (isAltPressed && isShiftPressed) {
+            return;
+        } else if (isAltPressed) {
+            delta *= this.props.minorStepSize;
+        } else if (isShiftPressed) {
+            delta *= this.props.majorStepSize;
+        }
+
+        // pretend we're incrementing from 0 if curValue isn't defined
+        const curValue = this.state.value || "0";
+
+        // truncate floating-point result to avoid precision issues when adding
+        // binary-unfriendly deltas like 0.1
+        const newValue = (!this.isValueNumeric(curValue))
+            ? 0
+            : parseFloat((parseFloat(curValue) + delta).toFixed(2));
+
+        this.setState({ shouldSelectAfterUpdate : true, value: newValue.toString() });
+    }
+
+
+    private isValueNumeric(value: string) {
+        // checking if a string is numeric in Typescript is a big pain, because
+        // we can't simply toss a string parameter to isFinite. below is the
+        // essential approach that jQuery uses, which involves subtracting a
+        // parsed numeric value from the string representation of the value. we
+        // need to cast the value to the `any` type to allow this operation
+        // between dissimilar types.
+        return value != null && ((value as any) - parseFloat(value) + 1) >= 0;
```

Comment:
May not be too important, but the `!isNan(+value)` approach returns `true` for empty strings, strings of whitespace, and empty arrays (though I guess the type system prevents empty arrays from seeping through in the first place). The approach above handles all those correctly. Might be worth taking the simpler approach though, if only for clarity.

I'll look into typeguards. My goal is to keep the value stored as a string (still need to tweak a couple things to make that consistently true)—will typeguards be able to treat stringified numbers as numbers? It's not clear that's the case from a quick read of the docs.